src/api.js

```diff
@@ -0,0 +1,159 @@
+let _ = require('lodash');
+
+let builder = new APIBuilder({
+  title: 'AMQP Messages API Documentation',
+  description: [
+    'This service is responsible for making pulse messages accessible',
+    'from browsers and cli. There are API endpoints to',
+    'bind / unbind to an exchange and pause / resume listening from a queue',
+  ].join('\n'),
+  projectName: 'taskcluster-evnets',
+  serviceName: 'events',
+  version: 'v1',
+  context: ['connection'],
+});
+
+
+// Returns JSON.parse(req.query.bindings) if everything goes well
+//   {"bindings" : [ 
+//     {"exchange" :  "a/b/c", "routingKey" : "a.b.c"},
+//     {"exchange" :  "x/y/z", "routingKey" : "x.y.z"},
+//   ]};
+var validateBindings = function(bindings) {
+  return JSON.parse(bindings);
+}
+
+builder.declare({
+  method: 'get',
+  route: '/connect/',
+  query: {
+    bindings: /./,
+  },
+  name: 'connect',
+  description: 'Connect to receive messages',
+  stability: APIBuilder.stability.experimental,
+  // Add input validation yml
+  title: 'Events-Api',
+}, async function(req, res) {
+  debug('hello');
+
+  // parse and validate 
+  var json_bindings = validateBindings(req.query.bindings);
+
+  // json_bindings.bindings contains array of {exchange, routingKey}
+  if (!json_bindings) {
+    // TODO :  Send error event through sendEvent and close connection.
+    return res.reportError('InvalidRequestArguments', "The bindings are not in specified json format");
+  }
+  debug('..bindings', json_bindings);
+
+  let abort;
+  const aborted = new Promise((resolve, reject) => abort = reject);
+  debug(aborted);
+
+  req.on('close', (err)=> {debug('aborting');abort(err);});
+
+  const sendEvent = (kind, data) => {
+    try {
+      var event = ['event: ' + kind,
+        'data: ' + JSON.stringify(data),
+        '\n',
+      ].join('\n');
+      //debug("sendEvent: ", event);
+      res.write(event);
+      debug('.....res.finished', aborted); 
+    } catch (err) {
+      debug('Error in sendEvent:');
+    }
+  };
+
+  try {
+
+    res.writeHead(200, {
+      'Content-Type': 'text/event-stream',
+      'Cache-Control': 'no-cache',
+    });
+    headWritten = true;
+
+    // TODO : add listener = PulseListener
+    var listener = new taskcluster.PulseListener({
+      prefetch:   5,
+      connection: this.connection,
+      maxLength:  50,
+    });
+    
+    _.forEach(json_bindings.bindings, (entry)=> listener.bind({
+      exchange: entry.exchange,
+      routingKeyPattern: entry.routingKey,
+    }));
+
+    // console.log(listener.connect());
+    // console.log(listener.resume());
+    
+    listener.connect().then(function() {
+      console.log('connected');
+      return listener.resume()
+    }, function(err) {
+      debug("..connect() error", err);
+      abort(err);
+    }).then(() =>
+      // The headers are written without any errors.
+      // This means we are ready to send messages.
+      sendEvent('ready', {}));
+
+
+    listener.on('message', (message)=> {
```

Comment:
You should set the listener before you connect, and you don't need to call `connect` just do `await listener.resume()`

Reply:
How will adding `await` help here?
The listener anyway wont receive messages before the `resume()` is finished

Reply:
True, nevermind